concurrency/MpmcQueue.h

```diff
@@ -5,112 +5,164 @@
 #include <cassert>
 namespace concurrency
 {
-// It is wise to make your queue entries at least 56B large
-// to facilitate cache line separation
+
+// High performing thread safe Multiple Producer, Multiple Consumer, Wait-free queue but...
+// If a reader is pre-empted when reading the element and a writer catches up, the writer will fail to push as the write
+// slot is still being read. This happens typically when the queue is rather full and there are competing threads in the
+// system.
 template <typename T>
 class MpmcQueue
 {
-    enum CellState : int
+    struct CellState
     {
-        emptySlot,
-        committed
+        uint32_t version = 0;
+        enum State : uint32_t
+        {
+            emptySlot = 0,
+            committed
+        } state = emptySlot;
     };
+
     struct Entry
     {
-        Entry() : state(emptySlot) {}
+        Entry() {}
 
+        T& value() { return reinterpret_cast<T&>(data); }
         std::atomic<CellState> state;
-        T value;
+        uint8_t data[sizeof(T)];
```

Comment:
32bits in front and no alignment restriction  on `data`. Seems safe for `T` with alignment requirement <= 32bits